### ISSUE TITLE
AWS Rekognition CompareFaces show high degree of similarity

### ISSUE THREAD

**Issue by xirkus** [Shawn-Shan/fawkes#125]
When taking two images (the original vs. the cloaked version) and running them through AWS Rekognition, the results returned show a high degree of similarity (99.8+%). This is unexpected given the following statement in your paper:

```
Our cloaking technique is not designed to fool a tracker who uses similarity matching. However, we believe our cloaking technique should still be effective against Amazon Rekognition, since cloaks create a feature space separation between original and cloaked images that should result in low similarity scores between them.
```

This results in images that are easy to match with their originals. Is this expected? Images were generated using fawkes 1.0.2 with the default settings. As a second experiment, i took two _different_ images to compare and still found that Rekognition was able to identify the unprotected image using the cloaked image.

**Comment by xirkus** [comment]
Testing with Azure Face, I got the following results:

1. original vs. cloaked (+95% similarity)
2. cloaked vs. different image (+87% similarity)

**Comment by xirkus** [comment]
Perhaps something a little more rigorous. Results for fawkes 0.3.2 and 1.0.2 (not that it makes a difference)

**TL;DR - I ran some experiments and found that fawkes was _unable_ to protect any images from identification. **

Read below if you'd like to reproduce for yourself. I was going to run tests for face++, but give the abysmal results, I'm convinced that fawkes does nothing to protect images from image classification.

machine specs:

- 3 GHz 6-Core Intel Core i5
- MacOS 11.3
- RAM 64Gb 2667 MHz DDR4

image specs:

- 337156 bytes
- dimensions 1150×1150  
- resolution 100x100
- colour space RGB

methodology:

- Using a single original image, run `fawkes` with the following settings (changing the location of the directory to suit your set up).
- AWS Rekognition allows you to use `CompareFaces` to determine if two images match based on similarity/confidence intervals. The similarity score is shown in the table below. Note that we used the default `NONE` QuailtyFilter for this query. 
- You can access MS Face Verification here: [Face](https://azure.microsoft.com/en-us/services/cognitive-services/face/). If you select the `Face verification` tab in the middle of the page you can upload your images for analysis/comparison.

fawkes 0.3.2

| setting | steps  | cost(s) | faces  | size(b) | rekognition | azure | face++ |
| ------- | -----  | ------- | -----  | ---- |----------- | ----- | ------ |
| `-d ./orig` |  25  |  16  | 1 | 743867 | 99.99998474121094 | 0.97847 | |
| `-m low -d ./orig` | 25 | 16 | 1 | 743867 | 99.99998474121094 | 0.97847 | |
| `-m mid -d ./orig ` | 50 | 76 | 1 | 762417 | 99.99667358398438 | 0.95172 | |
| `-m high -d ./orig` | 100 | 160 | 1 | 769281 | 99.9989242553711 | 0.94902 | |
| `-d ./orig --format jpeg` | 25 | 16 | 1 | 65433 | 99.99998474121094 | 0.97795 | |
| `-d ./orig --format jpeg -m low` | 25 | 16 | 1 | 65433 | 99.99998474121094 | 0.97795 | |
| `-d ./orig --format jpeg -m mid` | 50 | 77 | 1 | 67389 | 99.9981689453125 | 0.95443 | |
| `-d ./orig --format jpeg -m high` | 100 | 153 | 1 | 66652 | 99.99880981445312 | 0.94583| |

fawkes 1.0.2

| setting | steps  | cost(s) | faces  | size | rekognition | azure | face++ |
| ------- | -----  | ------- | -----  | ---- | ----------- | ----- | ------ |
| `-d ./orig` |  25  |  16  | 1 | 743867 | 99.99998474121094 | 0.97847 | |
| `-m low -d ./orig` | 25 | 16 | 1 | 743867 | 99.99998474121094 | 0.97847| |
| `-m mid -d ./orig ` | 50 | 80 | 1 | 762417| 99.99667358398438 | 0.95172| |
| `-m high -d ./orig` | 100 | 154 | 1 | 769281| 99.9989242553711 | 0.94902| |
| `-d ./orig --format jpeg` | 25 | 16 | 1 | 65433| 99.99998474121094 | 0.97795 | |
| `-d ./orig --format jpeg -m low` | 25 | 16 | 1 | 65433 | 99.99998474121094 | 0.97795 | |
| `-d ./orig --format jpeg -m mid` | 50 | 78 | 1 | 66652 | 99.9981689453125 | 0.95443 | |
| `-d ./orig --format jpeg -m high` | 100 | 150 | 1 | 67389 | 99.99880981445312 | 0.94583| |

Funny enough, I also read this the other day - [Legal chatbot firm DoNotPay adds anti-facial recognition filters to its suite of handy tools](https://www.theverge.com/2021/4/27/22405570/donotpay-ninja-anti-reverse-image-search-facial-recognition-filter). **Again, more snake oil.** All you have to do to reproduce their results against Google/Tineye is to flip the image horizontally.

Disappointed to say the least.

**Comment by Shawn-Shan** [comment]
Hi Mel, Thanks for the detailed testing. You probably did something wrong as v1.0 and v0.3 is very very different, so there is no way they achieves **identical** results. It is also important to test on standardized dataset. We benchmarked v1.0 high mode on PubFig dataset and it achieves 93+-9% protection success. 

I would love to help you take a look into what you did wrong in your evaluation.
